src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+                    .expect("failed to acquire lock")
+                    .free_io_addresses(res.addr.unwrap(), res.size),
+                IoType::PhysicalMmio | IoType::Mmio => self
+                    .resource
+                    .lock()
+                    .expect("failed to acquire lock")
+                    .free_mmio_addresses(res.addr.unwrap(), res.size),
+            }
+        }
+        self.resource
+            .lock()
+            .expect("failed to acquire lock")
+            .free_instance_id(id);
+    }
+
+    fn register_resources(
+        &mut self,
+        dev: Arc<Mutex<dyn Device>>,
+        resources: &mut Vec<IoResource>,
+    ) -> Result<()> {
+        for res in resources.iter() {
+            match res.res_type {
+                IoType::Pio => {
+                    if self
+                        .pio_bus
+                        .insert(Range(res.addr.unwrap(), res.size), dev.clone())
+                        .is_some()
+                    {
+                        return Err(Error::Overlap);
+                    }
+                }
+                IoType::Mmio => {
+                    if self
+                        .mmio_bus
+                        .insert(Range(res.addr.unwrap(), res.size), dev.clone())
+                        .is_some()
+                    {
+                        return Err(Error::Overlap);
+                    }
+                }
+                IoType::PhysicalMmio => continue,
+            };
+        }
+        Ok(())
+    }
+
+    fn allocate_irq_resource(
+        &mut self,
+        interrupt: Option<IrqResource>,
+    ) -> Result<Option<IrqResource>> {
+        match interrupt {
+            Some(IrqResource(irq)) => {
+                // Allocate irq resource
+                let irq_num = self
+                    .resource
+                    .lock()
+                    .expect("failed to acquire lock")
+                    .allocate_irq(irq)
+                    .map_err(Error::IrqAllocate)?;
+                Ok(Some(IrqResource(Some(irq_num))))
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn free_irq_resource(&mut self, interrupt: Option<IrqResource>) {
+        match interrupt {
+            Some(IrqResource(irq)) => self
+                .resource
+                .lock()
+                .expect("failed to acquire lock")
+                .free_irq(irq),
+            None => return,
+        }
+    }
+
+    /// Register a new device with its parent bus and resources request set.
+    pub fn register_device(
+        &mut self,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: &mut Vec<IoResource>,
+        interrupt: Option<IrqResource>,
+    ) -> Result<()> {
+        // Reserve resources
+        let id = self.allocate_resources(resources)?;
+
+        // Register device resources
+        if let Err(Error::Overlap) = self.register_resources(dev.clone(), resources) {
```

Comment:
Please do error recover too

Reply:
I guess you mean some operations here to let register succeed?